_posts/2022-08-16-zio-http.md

```diff
@@ -0,0 +1,778 @@
+official documentation
+[here](https://dream11.github.io/zio-http/docs/v1.x/dsl/server/) and
+[here](https://dream11.github.io/zio-http/docs/v1.x/examples/advanced-examples/advanced_server).
+
+## Next Steps
+
+At this stage, we've covered some `zhttp` basics, like pattern matching on a
+requests method and path, to run the appropriate logic. Out next steps will be
+about adding on extra functionality, like CRSF tokens and Authorization, via
+Middleware.
+
+### Middleware
+
+Broadly, the definition of _middleware_ is context dependent; in our realm of
+our discussion, if we're turning a `Request` into a `Response`, then it's
+anything we do _in the middle_ of that process. It may be something behind the
+scenes, like adding logging through the process, or even more intrusive like
+modifying the request to add headers to the request. Middleware is great at
+helping de-couple/re-use business logic. I use the word _intrusive_ to help
+emphasize that de-couple doesn't mean make _optional/not required_. For example,
+if we adding logging, we probably wouldn't expect that to have an impact on the
+functionality of our application. If we're parsing out information from a
+header, and then adding a custom, internal authorization header, then our
+application might stop working if we don't include it.
+
+Specificity, in the context of `zhttp`, a `Middleware` is a transformation
+function that converts _one_ `Http` to _another_.
+
+```scala
+type Middleware[R, E, AIn, BIn, AOut, BOut] = Http[R, E, AIn, BIn] => Http[R, E, AOut, BOut]
+```
+
+We _attach_ middleware to our `Http` via the `@@` operator. For example, we
+could update our logic to use a built-in debug `Middleware` like so:
+
+```scala
+  val wrapped: Http[Console with Clock, IOException, Request, Response] =
+    combined @@ Middleware.debug
+```
+
+and then, when running our application, we would see some debug messaging
+printed when a client interacts with our server:
+
+```shell
+[info] Starting server on http://localhost:9000
+[info] 200 GET /owls 9ms
+```
+
+In the next section, we will build our own Logging Middleware, and then look at
+a few of the other built in ones.
+
+#### Logging
+
+Our custom middleware is going to log some information about the `Request`
+received, and the `Response` about to be sent back. We'll set up a new object
+`Verbose` and define a method `log` that returns a `new Middleware`, in which we
+will define the trait's `apply` method.
+
+```scala
+package com.alterationx10.troto.middleware
+
+import zhttp.http._
+import zio._
+
+object Verbose {
+
+  def log[R, E >: Throwable]
+      : Middleware[R, E, Request, Response, Request, Response] =
+    new Middleware[R, E, Request, Response, Request, Response] {
+
+      override def apply[R1 <: R, E1 >: E](
+          http: Http[R1, E1, Request, Response]
+      ): Http[R1, E1, Request, Response] =
+        http
+          .contramapZIO[R1, E1, Request] { r =>
```

Comment:
I think we should add examples of `.mapZIO`/`.contramapZIO` to the transformers next to the http combinators